### ISSUE TITLE
Video streaming directly from GPU source?

### ISSUE THREAD

**Issue by simonmeister** [WonderMediaProductions/webrtc-dotnet-core#1]
Hi,

thanks for your code! I am reading through it (mainly the WebRTC-native C++ directory) to understand more about using native webrtc for p2p video streaming. One question beforehand, regarding the YUV handling. As far as I understood, webrtc uses YUV420 natively. I have a D3D texture in GPU memory with BGRA32 format and would like to encode it with NvPipe, stream via WebRTC, and reconstruct it as a GPU texture on the receiver side. Just so that I understand it correctly, the adaptions in this code seems to enable streaming without using any explicit YUV420 conversions or GPU/CPU memory copies. Is this correct?

Thanks
Simon

**Comment by ziriax** [comment]
Hi Simon,

First of all, unlike other frameworks like Microsoft's 3D streaming toolkit, we didn't make any adaptations to Google's WebRTC native itself, we are using a recent master version.

And yes, if you have an Nvidia card that supports nvenc and has a very recent driver (as of last month), the encoding should be done completely on the GPU-card (actually by another ASIC on the GPU that runs completely in parallel with the GPU itself). No copying should be done, at least that is what NVidia told me. It is still possible that some YUV conversion is done before passing the texture to the NVENC chip, I don't know yet, we are looking into this (we will have our shader output YUV directly in this case).

Now it seems that the CPU is still used by the OS, and not in the process that does the encoding, this might be driver overhead, not sure.

Since my employer only uses D3D11 and Windows, we might modify this code to accept pre-encoded H264 data using the NVENC SDK directly, instead of going through the CUDA API of NvPipe, because CUDA could introduce some overhead, and we don't need that really.

Note that this library is not finished at all, currently only supports Windows and D3D11, and the API will change, but nevertheless we found it useful to share the code, because it is a real pain to get this working.

Especially the video format enumeration is really bad, it actually combines both the pixel format as the encoder that must be used (RGBA32 means a CPU bitmap that will be encoded using VP8 software, and that currently only works when no NVENC is found or was disabled), and CpuTexture and GpuTexture means H264 hardware encoding using NVENC... We want to split this up, but I haven't found a way to do this yet (WebRTC native has no docs at all, besides a very good source browser, some samples, and a forum that never replied to my questions ;-))

**Comment by simonmeister** [comment]
Thanks! Do you also support decoding into a GPU texture? I wasn't able to find any starting points here regarding that.

**Comment by ziriax** [comment]
No, unfortunately, we don't need that yet.
